net: LwM2M: Notify Timeout Handling

JPSELC opened the pull request:
Added a new LWM2M_ENGINE_NOTIFICATION_TIMEOUT_REGISTRATION_UPDATE
 Kconfig option to enable automatic registration update when
 notify timeout occurs

Added function type definition for external notification of
 notify timeout : ext_notify_timeout_cb_t

Added function lwm2m_engine_set_ext_notify_timeout_cb_fn to set
 the ext notify function

When notify message timeout occurs, a call is made to
 ext_notify_timeout_cb if it exists,
 a call of engine_trigger_update is made if enabled,
 and an error message is issued.

Signed-off-by: John Power <john.power@xylem.com>

JPSELC committed:
net: LwM2M: Notify Timeout Handling

Added a new LWM2M_ENGINE_NOTIFICATION_TIMEOUT_REGISTRATION_UPDATE
 Kconfig option to enable automatic registration update when
 notify timeout occurs

Added function type definition for external notification of
 notify timeout : ext_notify_timeout_cb_t

Added function lwm2m_engine_set_ext_notify_timeout_cb_fn to set
 the ext notify function

When notify message timeout occurs, a call is made to
 ext_notify_timeout_cb if it exists,
 a call of engine_trigger_update is made if enabled,
 and an error message is issued.

Signed-off-by: John Power <john.power@xylem.com>

JPSELC committed:
net: lwm2m: lwm2m_notify_timeout_handling
Added notify_timeout_cb to struct lwm2m_ctx to allow application to
 handle notify timeout
Added lwm2m_rd_client_update to lwm2m.h to allow application to
 trigger registration update
Added notify_message_timeout_cb which calls notify_timeout_cb from
 struct lwm2m_ctx and logs an error message

Signed-off-by: John Power <john.power@xylem.com>

carlescufi commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -87,6 +87,11 @@ config LWM2M_ENGINE_MAX_OBSERVER
 	  This value sets the maximum number of resources which can be
 	  added to the observe notification list.
 
+config LWM2M_ENGINE_NOTIFICATION_TIMEOUT_REGISTRATION_UPDATE
```
Thanks for the PR! Could you please shorten this option name? This is way too long

JPSELC replied:
LWM2M_ENGINE_TIMEOUT_REG_UPDATE okay?

JPSELC committed:
net: LwM2M: Minor Fixes

Reduced length of Kconfig option name
Fixed space required before the open brace

Signed-off-by: John Power <john.power@xylem.com>

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4247,6 +4250,17 @@ static int32_t retransmit_request(struct lwm2m_ctx *client_ctx,
+		ext_notify_timeout_cb(msg);
+	}
+#if defined(CONFIG_LWM2M_ENGINE_TIMEOUT_REG_UPDATE)
+	engine_trigger_update(false);
```
I just wonder, this mechanism isn't covered by the spec in any way I believe? Perhaps, instead of doing this automatically, we should allow the application to deal with the notification updates on its own? So instead calling app callback and sending the trigger for update automatically, provide a new API function, something like `lwm2m_rd_client_update()`, to trigger the update in the LwM2M state machine?

JPSELC replied:
Although this isn't part of the lwm2m specification, I think it would be a useful KConfig option (which is disabled by default), as it fits well with what is proposed for https://github.com/zephyrproject-rtos/zephyr/issues/32610
I can however do as you suggest, and leave it to the application to trigger the update

rlubos replied:
I think that leaving the decision to the application what to do in case of notification timeout gives more flexibility for the user. We see that one solution is to trigger the Register Update, so that the engine can detect connection issues on its own, but some may want to act more actively and restart the registration process right away. 

Generally, I think that within LwM2M engine we should stick as close to the specification as possible - if some non-standard action can be easily achieved from the application (as it is in this scenario), I strongly opt for such a solution.

rlubos commented:
And a general comment - please use lowercase `net: lwm2m: ` prefix for the commit messages, it makes grepping easier. And please squash any commits with review fixes into the original one.

JPSELC commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -138,6 +138,9 @@ static struct lwm2m_engine_obj *get_engine_obj(int obj_id);
 							 int obj_inst_id);
 
+/* function pointer to indicate notification timeout to the application */
+static ext_notify_timeout_cb_t ext_notify_timeout_cb;
```
sm_do_registration calls lwm2m_engine_start(client.ctx) in lwm2m_rd_client.c which is the first use of the lwm2m_ctx structure. If notify_timeout_cb is part of lwm2m_ctx, it should really be set in sm_do_registration, but I don't know how best to do that by the application.
 I guess we can set it later using a lwm2m_engine_set_ext_notify_timeout_cb function? Also I'm not sure if that callback can take any useful parameters, so it will be just a general notification that a timeout has occurred

JPSELC replied:
I think the solution is add notify_timeout_cb to lwm2m_ctx and set it before the application calls lwm2m_rd_client_start()

rlubos replied:
Yes, we already do this with `tls_tag` for instance. IMO keeping thigs like this within `lwm2m_ctx` structure is more future-proof, in case we finally support connection to multiple LwM2M servers (not sure when will this happen though).

JPSELC committed:
Merge branch 'zephyrproject-rtos:main' into lwm2m_notify_timeout_handling

JPSELC committed:
Merge branch 'lwm2m_notify_timeout_handling' of https://github.com/JPSELC/zephyr into lwm2m_notify_timeout_handling

JPSELC committed:
Merge branch 'lwm2m_notify_timeout_handling' of https://github.com/JPSELC/zephyr into lwm2m_notify_timeout_handling
